src/libpgmoneta/utils.c

```diff
@@ -3003,6 +3004,198 @@ pgmoneta_is_file_archive(char* file_path)
    return false;
 }
 
+int
+pgmoneta_token_bucket_init(struct token_bucket* tb, unsigned long burst, unsigned long max_rate, int every)
+{
+   if (tb != NULL)
+   {
+      tb->burst = burst;
+      atomic_init(&tb->cur_tokens, burst);
+      tb->max_rate = max_rate;
+      tb->every = every;
+      atomic_init(&tb->last_time, (unsigned long)time(NULL));
+      return 0;
+   }
+   return 1;
+}
+
+void
+pgmoneta_token_bucket_destroy(struct token_bucket* tb)
+{
+   if (tb)
+   {
+      free(tb);
+   }
+}
+
+int
+pgmoneta_token_bucket_add(struct token_bucket* tb)
+{
+   unsigned long diff;
+   unsigned long expected_tokens;
+   unsigned long new_tokens;
+   unsigned long expected_time;
+   unsigned long cur_time;
+
+   diff = (unsigned long) time(NULL) - atomic_load(&tb->last_time);
+
+   if (diff <= 0)
+   {
+      return 0;
+   }
+
+   expected_tokens = atomic_load(&tb->cur_tokens);
+   new_tokens = expected_tokens + tb->max_rate * diff / tb->every;
+
+   if (new_tokens > tb->burst)
+   {
+      new_tokens = tb->burst;
+   }
+
+   while (!atomic_compare_exchange_weak(&tb->cur_tokens, &expected_tokens, new_tokens))
+   {
+      expected_tokens = atomic_load(&tb->cur_tokens);
+      new_tokens = expected_tokens + tb->max_rate * diff / tb->every;
+
+      if (new_tokens > tb->burst)
+      {
+         new_tokens = tb->burst;
+      }
+   }
+
+   expected_time = atomic_load(&tb->last_time);
+   cur_time = (unsigned long) time(NULL);
+   while (!atomic_compare_exchange_weak(&tb->last_time, &expected_time, cur_time))
```

Comment:
Correct me if I'm wrong, but think of this situation: suppose you add 1 token/sec, and last_time=1, at time=2 a thread comes asking the bucket, so that one token is added to the bucket. But before you update last_time, another thread at time=3 comes asking the bucket, since last_time is still 1, 2 tokens are added. So tokens at time=2 are added twice.

Reply:
Yes, you are right. I will make them update atomically or other ways.

Reply:
`
int 
pgmoneta_token_bucket_add(struct token_bucket* tb)
{
   unsigned long diff;
   unsigned long expected_tokens;
   unsigned long new_tokens;
   unsigned long expected_time;
   unsigned long cur_time;

   cur_time = (unsigned long) time(NULL);
   expected_time = atomic_load(&tb->last_time);

   while (!atomic_compare_exchange_weak(&tb->last_time, &expected_time, cur_time))
   {
      expected_time = atomic_load(&tb->last_time);
	  cur_time = (unsigned long) time(NULL);
   }

   diff = cur_time - expected_time;

   if (diff < tb->every)
   {
      return 0;
   }

   expected_tokens = atomic_load(&tb->cur_tokens);
   new_tokens = expected_tokens + tb->max_rate * diff / tb->every;

   if (new_tokens > tb->burst)
   {
      new_tokens = tb->burst;
   }

   while (!atomic_compare_exchange_weak(&tb->cur_tokens, &expected_tokens, new_tokens))
   {
      expected_tokens = atomic_load(&tb->cur_tokens);
      new_tokens = expected_tokens + tb->max_rate * diff / tb->every;

      if (new_tokens > tb->burst)
      {
         new_tokens = tb->burst;
      }
   }

   return 0;
}`

Do you think this approach is better? By switching the order, even if two threads try to update the token, they will still update their respective sections based on their time differences. This should not affect the result. I think this will address the race condition.

Reply:
int
    pgmoneta_token_bucket_add(struct token_bucket* tb)
    {
        unsigned long diff;
        unsigned long expected_tokens;
        unsigned long new_tokens;
        unsigned long expected_time;
        unsigned long cur_time;

        cur_time = (unsigned long) time(NULL);
        expected_time = atomic_load(&tb->last_time);

        while (!atomic_compare_exchange_weak(&tb->last_time, &expected_time, cur_time))
        {
            expected_time = atomic_load(&tb->last_time);
            cur_time = (unsigned long) time(NULL);
        }

        diff = cur_time - expected_time;

        if (diff < tb->every)
        {
            return 0;
        }

        expected_tokens = atomic_load(&tb->cur_tokens);
        new_tokens = expected_tokens + tb->max_rate * diff / tb->every;

        if (new_tokens > tb->burst)
        {
            new_tokens = tb->burst;
        }

        while (!atomic_compare_exchange_weak(&tb->cur_tokens, &expected_tokens, new_tokens))
        {
            expected_tokens = atomic_load(&tb->cur_tokens);
            new_tokens = expected_tokens + tb->max_rate * diff / tb->every;

            if (new_tokens > tb->burst)
            {
                new_tokens = tb->burst;
            }
        }

        return 0;
    }

Do you think this approach is better? By switching the order, even if two threads try to update the token, they will still update their respective sections based on their time differences. This should not affect the result. I think this will address the race condition.

Reply:
Another good idea to avoid overflow is to get the time it takes to fill up a bucket, if diff is longer than that, just fill up the bucket directly - got this idea from Go's built-in rate limiter.

Reply:
How about this way, we calculate the `diff` before the update. This way, the current time will not be updated until it is greater than the specified `every` interval. And, in the while update time loop, we continue to track the `diff`. If the `diff` is below the 'every', we exit the function.

Reply:
That might work~ Note that we have a built-in worker pool, and it's currently used in [wf_restore.c](https://github.com/pgmoneta/pgmoneta/blob/main/src/libpgmoneta/wf_restore.c), could be a good idea to configure workers and add your rate limiter there and test it out, i.e. restore a backup.

Another thing that worries me is the granularity - if your data buffer to write to disk is larger than the bucket capacity, you risk looping forever. It's probably worthwhile to add a wrapper function with rate limit around fwrite, where you can partially write a buffer each time until it has all been written to disk. But this is open for discussion.

Reply:
> if your data buffer to write to disk is larger than the bucket capacity, you risk looping forever

I noticed this issue, so I added a line to exit the backup if this situation is encountered and to print the error. Normally, we need the burst to be larger than the buffer, as the burst will not harm the rate limit; it's just a one-time event. We limit it after the burst is consumed. Therefore, even if the burst is set to a large number, it will not affect our rate limit.

Reply:
I should set a limit [here](https://github.com/pgmoneta/pgmoneta/blob/ca773ca4c088909db7ced2ded3305085a865d3b9/src/libpgmoneta/utils.c#L1875-L1895) to limit restore, right? What do you mean by 'configure workers'? If we simply set the limit here, shouldn't it be limited?

Reply:
> I should set a limit [here](https://github.com/pgmoneta/pgmoneta/blob/ca773ca4c088909db7ced2ded3305085a865d3b9/src/libpgmoneta/utils.c#L1875-L1895) to limit restore, right?

I think so.

> What do you mean by 'configure workers'? If we simply set the limit here, shouldn't it be limited?
Sorry about the confusion, it's simply for testing - if you want see if your bucket works well in multi-threading environment, you can test it out there, with [workers configured](https://github.com/pgmoneta/pgmoneta/blob/main/doc/CONFIGURATION.md). A worker pool will be created to copy files with a shared token bucket.

Reply:
Can we simply place the token bucket in shared memory so that any process needing rate limiting can access and use it? However, this approach would allow only one rate limit configuration to be used universally.

Reply:
It's open for discussion. I would prefer one rate limiter per process, either it's for backup streaming or restoring or some other task. However, it can be shared among the worker threads within the process. Have you encountered any trouble with this?

Reply:
I am currently testing the restore process, but when I run `pgmoneta-cli restore primary newest /tmp/restore`, the log shows `2024-03-28 19:53:02 INFO restore.c:106 Restore: primary/20240328151050 (Elapsed: 00:00:00)`. However, I can't find the restored files in `/tmp/restore`. This issue is unrelated to the `WAL segment error`, right?

Reply:
Hmm, odd, I just restored a backup, and it works fine... at least the data is there. And what do you mean by `WAL segment error`?

Reply:
Have you checked the data in your backup, is it there in the first place?

Reply:
> And what do you mean by WAL segment error?

There is alway an error about the WAL 
`2024-03-30 08:34:08 INFO  configuration.c:1466 No WAL slot defined for primary`

Reply:
![image](https://github.com/pgmoneta/pgmoneta/assets/113565433/ccf83856-0f7b-42e5-b01c-d01b9c59415b)

Does this mean that the restore was successful? I just tested it again today, and it worked, which is so weird. I believe there must be something wrong somewhere.

Reply:
Did you define `wal_slot = ...` and `create_slot = yes` ?

If there are questions during the startup process that isn't covered by the documentation then we need to fix those

Reply:
I configured `wal_level = replica` in `postgresql.conf`, but what does `create_slot = yes` mean? Do I need to configure it in `pgmoneta.conf`?

Reply:
> but what does create_slot = yes

It's for creating a replication slot for WAL streaming. You see WAL segments are generated endlessly, the server cleans up outdated segments regularly. A slot lets server know how much segments have you saved on your end, so that it doesn't clean up WAL files you haven't yet received.

If you don't configure it, you are fine if you keep up. But if you fall behind, say you stopped pgmoneta for a while while the server is still running, you risk losing some segments for server has deleted them.

Reply:
`pgmoneta_management_read_int32: read: 4 Invalid argument`
What does the management do? When I set the workers and token, the restore process finished at this line. However, when I don't use workers, this problem does not occur.

Reply:
When I make the token shared and use it to limit the restore process in [copy_file](https://github.com/pgmoneta/pgmoneta/blob/9dfbbb9421577ecbc0e5c2952442b67d44efd26d/src/libpgmoneta/utils.c#L1878-L1953), the function `pgmoneta_management_read_int32` quits the process before the restoration begains. However, if I don't use the token, everything works well.

Reply:
`pgmoneta_management_read_int32` is for cli, when you issue restore command, cli sends it to pgmoneta, then tries to read the restoring outcome status from pgmoneta. So some error occurred in between, either pgmoneta fails to send the result, or the connection broke.

Reply:
Try adding more loggs to see what happens in between.

Reply:
> pgmoneta_management_read_int32 is for cli, when you issue restore command, cli sends it to pgmoneta, then tries to read the restoring outcome status from pgmoneta. So some error occurred in between, either pgmoneta fails to send the result, or the connection broke.

See [restore()](https://github.com/pgmoneta/pgmoneta/blob/9dfbbb9421577ecbc0e5c2952442b67d44efd26d/src/cli.c#L923)

Reply:
Chances are rate limit slows down the process, and [read_complete()](https://github.com/pgmoneta/pgmoneta/blob/9dfbbb9421577ecbc0e5c2952442b67d44efd26d/src/libpgmoneta/management.c#L1510) times out. Try increasing number of retries to see if problem persists.

Reply:
I believe the token bucket can function in a multi-threaded environment and perform rate limiting. I tested it with `works=2` and the restore was successful. Below is the log, and I didn't find any issues there.
[pgmoneta.log](https://github.com/pgmoneta/pgmoneta/files/14813354/pgmoneta.log)

Reply:
Did you solve your previous problem? What's the issue there?

Reply:
I made a mistake in the shared memory, so the management can't receive the response from the restore process. So, it resulted in an error.

Reply:
How do we handle the token bucket now? Should we set different ones for different functions or just use a shared one?

Reply:
I would still prefer one bucket per process. Jesper may think otherwise. He'll join us later :)

Reply:
One bucket per process